File: java-bdd-template/src/test/java/com/example/steps/Steps.java
package com.example.steps;

import io.cucumber.java.en.*;
import static org.junit.Assert.*;
import java.util.*;

public class Steps {
  private Map<String,String> users = new HashMap<>();
  private List<String> catalog = new ArrayList<>();
  private boolean dashboardVisible;
  private String lastError;
  private String query;

  @Given("existe una usuaria {string} con contraseña {string}")
  public void seedUser(String email, String pwd){
    users.put(email, pwd);
  }

  @When("inicio sesión con email {string} y contraseña {string}")
  public void login(String email, String pwd){
    if(!users.containsKey(email)) lastError="Usuario no existe";
    else if(!Objects.equals(users.get(email), pwd)) lastError="Credenciales inválidas";
    else dashboardVisible=true;
  }

  @Then("debo ver el panel principal")
  public void seeDashboard(){
    assertTrue(dashboardVisible);
  }

  @Then("debo ver el error {string}")
  public void seeError(String msg){
    assertEquals(msg, lastError);
  }

  @Given("el catálogo contiene:")
  public void seedCatalog(io.cucumber.datatable.DataTable table){
    catalog.clear();
    List<Map<String,String>> rows = table.asMaps(String.class, String.class);
    for(Map<String,String> row : rows){
      catalog.add(row.get("nombre"));
    }
  }

  @When("busco {string}")
  public void search(String q){
    this.query = q;
  }

  @Then("los resultados deben incluir {string}")
  public void resultsInclude(String name){
    assertTrue(catalog.contains(name));
  }

  @Then("veo el mensaje {string}")
  public void seeNoResults(String msg){
    if(query == null || !catalog.contains(query)){
      assertEquals("No se encontraron resultados", msg);
    } else {
      fail("Había resultados, no debería mostrarse el mensaje de 'No se encontraron resultados'.");
    }
  }
}

File: python-behave-template/features/steps/steps.py
from behave import given, when, then
from hamcrest import assert_that, equal_to

users = {}
catalog = []
dashboard_visible = False
last_error = None
query = None

@given('existe una usuaria "{email}" con contraseña "{pwd}"')
def step_impl(context, email, pwd):
    users[email] = pwd

@when('inicio sesión con email "{email}" y contraseña "{pwd}"')
def step_impl(context, email, pwd):
    global dashboard_visible, last_error
    if email not in users:
        last_error = "Usuario no existe"
    elif users[email] != pwd:
        last_error = "Credenciales inválidas"
    else:
        dashboard_visible = True

@then("debo ver el panel principal")
def step_impl(context):
    assert dashboard_visible is True

@then('debo ver el error "{msg}"')
def step_impl(context, msg):
    assert_that(last_error, equal_to(msg))

@given("el catálogo contiene:")
def step_impl(context):
    global catalog
    # DataTable con encabezado 'nombre'
    catalog = [row["nombre"] for row in context.table]

@when('busco "{q}"')
def step_impl(context, q):
    global query
    query = q

@then('los resultados deben incluir "{name}"')
def step_impl(context, name):
    assert name in catalog

@then('veo el mensaje "No se encontraron resultados"')
def step_impl(context):
    if query not in catalog:
        shown = "No se encontraron resultados"
        assert "No se encontraron" in shown
    else:
        raise AssertionError("Se esperaba no encontrar resultados, pero sí había.")
